File: Edison/MPU9250/mpu.c
#include <stdio.h>
#include <unistd.h>
#include <mraa.h>

#define    MPU9250_ADDRESS            0x68
#define    MAG_ADDRESS                0x0C

#define    GYRO_FULL_SCALE_250_DPS    0x00  
#define    GYRO_FULL_SCALE_500_DPS    0x08
#define    GYRO_FULL_SCALE_1000_DPS   0x10
#define    GYRO_FULL_SCALE_2000_DPS   0x18

#define    ACC_FULL_SCALE_2_G        0x00  
#define    ACC_FULL_SCALE_4_G        0x08
#define    ACC_FULL_SCALE_8_G        0x10
#define    ACC_FULL_SCALE_16_G       0x18

int main()
{
	mraa_init();
	mraa_i2c_context mpu;
	mpu = mraa_i2c_init(0);
	mraa_i2c_address(mpu, MPU9250_ADDRESS);
	
	mraa_i2c_write_byte_data(mpu, 0x06, 29);
	mraa_i2c_write_byte_data(mpu, 0x06, 26);
	mraa_i2c_write_byte_data(mpu, ACC_FULL_SCALE_2_G, 28);
	mraa_i2c_write_byte_data(mpu, GYRO_FULL_SCALE_1000_DPS, 27);

	//////////////////////// write_of_slave
	mraa_i2c_write_byte_data(mpu, 32, 106);
	mraa_i2c_write_byte_data(mpu, 0x0C, 49);
	mraa_i2c_write_byte_data(mpu, 10, 50);
	mraa_i2c_write_byte_data(mpu, 22, 51);
	mraa_i2c_write_byte_data(mpu, 128, 52);
	//////////////////////// read_of_slave
	mraa_i2c_write_byte_data(mpu, 32, 106);
	mraa_i2c_write_byte_data(mpu, 140, 37);
	mraa_i2c_write_byte_data(mpu, 3, 38);
	mraa_i2c_write_byte_data(mpu, 136, 39);
	//////////////////////// read_of_slave
	mraa_i2c_write_byte_data(mpu, 18, 10);
	////////////////////////
	
	int sample = 0;
	while(sample<1000)
	{
		uint8_t Buf[14];
		mraa_i2c_read_bytes_data(mpu, 59, Buf, 14);
		// Accelerometer
		int16_t ax = -(Buf[0] << 8 | Buf[1]);
		int16_t ay = -(Buf[2] << 8 | Buf[3]);
		int16_t az = Buf[4] << 8 | Buf[5];
		// Gyroscope
		int16_t gx = -(Buf[8] << 8 | Buf[9]); 
		int16_t gy = -(Buf[10] << 8 | Buf[11]);
		int16_t gz = Buf[12] << 8 | Buf[13];
		// Magnetometer
		mraa_i2c_read_bytes_data(mpu, 73, Buf, 6);
		int16_t mx = -(Buf[1] << 8 | Buf[0])+195;
		int16_t my = -(Buf[3] << 8 | Buf[2])+120;
		int16_t mz = Buf[5] << 8 | Buf[4]-196;
		int result_ag[6] = {ax, ay, az, gx, gy, gz};
		int result_m[3] = {mx, my, mz};
		printf("%6d,%6d,%6d,%6d,%6d,%6d,%6d,%6d,%6d\n",ax,ay,az,gx,gy,gz,mx,my,mz);
   
	}
  return 0;
}

File: Video/Edison/main/include/mpu9250/old_code/mpu.py
import mraa
import time
import math
import numpy

raw_offset        =         32767
MPU9250_ADDRESS   =         0x68
MAG_ADDRESS       =         0x0C

GYRO_FULL_SCALE_250_DPS  =  0x00  
GYRO_FULL_SCALE_500_DPS  =  0x08
GYRO_FULL_SCALE_1000_DPS =  0x10
GYRO_FULL_SCALE_2000_DPS =  0x18

ACC_FULL_SCALE_2_G     =   0x00  
ACC_FULL_SCALE_4_G     =   0x08
ACC_FULL_SCALE_8_G     =   0x10
ACC_FULL_SCALE_16_G    =   0x18

mpu = mraa.I2c(0)
mag = mraa.I2c(0)

mpu.address(MPU9250_ADDRESS)
mag.address(MAG_ADDRESS)

mpu.writeReg(29,0x06)
mpu.writeReg(26,0x06)
mpu.writeReg(27,GYRO_FULL_SCALE_1000_DPS)
mpu.writeReg(28,ACC_FULL_SCALE_2_G)

############## write_of_slave
mpu.writeReg(106,32)
mpu.writeReg(49,0x0C)
mpu.writeReg(50,10)
mpu.writeReg(51,22)#mag mode
mpu.writeReg(52,128)
##############


############## read_of_slave
mpu.writeReg(106,32)
mpu.writeReg(37,140)
mpu.writeReg(38,3)
mpu.writeReg(39,136)
##############

mag.writeReg(10,18)



time.sleep(1)
buf = [0,0,0,0,0,0,0,0,0,0,0,0,0,0]
gx=[]
gy=[]
gz=[]
ax=[]
ay=[]
az=[]
mx=[]
my=[]
mz=[]

sample = 0
tmp = 0

def readmag():
    raw_mag = [0,2,0,0,0,0,0]
    if(mag.readReg(2) & 0x01):
        for i in range(0,7):
            raw_mag[i]=mag.readReg(3+i)
    return raw_mag
        
            

while sample<10000:
    for i in range(0,14):
        buf[i]=mpu.readReg(59+i)
        #time.sleep(0.001)
    tmp = buf[0]*256 + buf[1] + raw_offset
    if(tmp > 65536):
        tmp -= 65536
    ax.append(-tmp+raw_offset)
    tmp = buf[2]*256 + buf[3] + raw_offset
    if(tmp > 65536):
        tmp -= 65536
    ay.append(-tmp+raw_offset)
    tmp = buf[4]*256 + buf[5] + raw_offset
    if(tmp > 65536):
        tmp -= 65536
    az.append(tmp-raw_offset)
    
    tmp = buf[8]*256 + buf[9] + raw_offset
    if(tmp > 65536):
        tmp -= 65536
    gx.append(-tmp+raw_offset);   
    tmp = buf[10]*256 + buf[11] + raw_offset
    if(tmp > 65536):
        tmp -= 65536
    gy.append(-tmp+raw_offset);
    tmp = buf[12]*256 + buf[13] + raw_offset
    if(tmp > 65536):
        tmp -= 65536
    gz.append(tmp-raw_offset);
    
    
    for i in range(0,6):
        buf[i]=mpu.readReg(73+i)
    tmp = buf[1]*256 + buf[0] + raw_offset
    if(tmp > 65536):
        tmp -= 65536
    mx.append(-tmp+raw_offset+195)
    tmp = buf[3]*256 + buf[2] + raw_offset
    if(tmp > 65536):
        tmp -= 65536
    my.append(-tmp+raw_offset+120)
    tmp = buf[5]*256 + buf[4] + raw_offset
    if(tmp > 65536):
        tmp -= 65536
    mz.append(-tmp+raw_offset-196)
    
    result = [ax[sample],ay[sample],az[sample],gx[sample],gy[sample],gz[sample]]
    result = [mx[sample],my[sample],mz[sample]]
    test = result[0]*result[0]+result[1]*result[1]+result[2]*result[2]#there is an offset, so the value is not stable
    #print(mraa.getI2cBusId(6))
#    print(result)
#    print(gy)
#    print(gz)
#    test = (mpu.readReg(74)*256+mpu.readReg(73))+raw_offset
#    if(test > 65536):
#        test -= 65536
#    test -= raw_offset
#    test = mag.readReg(2)
#    test = readmag()
#    print(mag.readReg(0))
    print (test)
    sample += 1

output = [mx,my,mz]
f=open('demo.txt','w')
for i in output:
    k=' '.join([str(j) for j in i])
    f.write(k+"\n")
f.close()
